refactor!: Keyring migration

cyberbono3 opened the pull request:
<!--
The default pull request template is for types feat, fix, or refactor.
For other templates, add one of the following parameters to the url:
- template=docs.md
- template=other.md
-->

## Description

The draft PR #9222 
Closes: #7108

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review. -->

- implement proto definition for `Record` 
- rename `Info.go` to `legacyInfo.go` within `keyring` package
- implement CLI `migrate` command that migrates all keys from  legacyInfo to proto according to @robert-zaremba migration [algorithm](https://github.com/cosmos/cosmos-sdk/pull/9222/#discussion_r624683839)
- remove legacy keybase entirely.
- add `Migrate` and `MigrateAll` functions  in `keyring.go` for single key and all keys migration
- add tests
- fix tests

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [x] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

cyberbono3 commented on client/keys/delete.go:
```diff
@@ -53,7 +53,7 @@ private keys stored in a ledger device cannot be deleted with the CLI.
 				}
 
-				if info.GetType() == keyring.TypeLedger || info.GetType() == keyring.TypeOffline {
+				if k.GetType() == keyring.TypeLedger || k.GetType() == keyring.TypeOffline {
```
I would like to remove `GetType` func from `Record` entirely. I think, we don not need that, do we?

cyberbono3 commented on proto/cosmos/crypto/keyring/types.proto:
```diff
@@ -0,0 +1,36 @@
+syntax = "proto3";
+package cosmos.crypto.keyring.v1;
```
```
{"path":"proto/cosmos/crypto/keyring/types.proto","start_line":2,"start_column":1,"end_line":2,"end_column":34,"type":"PACKAGE_DIRECTORY_MATCH","message":"Files with package \"cosmos.crypto.keyring.v1\" must be within a directory \"cosmos/crypto/keyring/v1\" relative to root but were in directory \"cosmos/crypto/keyring\"."}
```

Since I added `v1` to package (proto convention), linter [suggest](url) to move `types.proto` from `*/keyring/.` to `*/keyring/v1` That action will bring additional refactoring work and some complexity since `isRecord_Item`  is private interface defined in `types.proto` and it is used in `record.go`. Therefore, I have to move `/keyring/record.go` and `/keyring/record.go` logic to `keyring/v1` package. Does it refactoring necessary or I can leave as it right now?

amaurym commented on .gitignore:
```diff
@@ -22,6 +22,8 @@ dist
 tools-stamp
 buf-stamp
 artifacts
+private/*.sh
+test
```
```suggestion
```

amaurym commented on proto/cosmos/crypto/keyring/types.proto:
```diff
@@ -0,0 +1,36 @@
+  oneof item {
+    Local local = 3;
+    Ledger ledger  = 4;
+    Empty empty  = 5;
```
Why are we using empty? I think using `Multi` and `Offline` would be clearer, even both of them are empty structs

amaurym replied:
Also, we should add some comments for each of the enum types

cyberbono3 replied:
> Why are we using empty? I think using `Multi` and `Offline` would be clearer, even both of them are empty structs

@robert-zaremba suggested to use `Empty` instead of `Multi` and `Offline`.

cyberbono3 replied:
@amaurym Following your comment, I went ahead and added `Multi` and `Offline` as empty structs . You can review my code [here](https://github.com/cosmos/cosmos-sdk/pull/9695/commits/60b8df9c26668a5b4766bd73addb7a4f3fd1951f)

amaurym commented on proto/cosmos/crypto/hd/hd.proto:
```diff
@@ -0,0 +1,18 @@
+syntax = "proto3";
+package cosmos.crypto.hd;
```
```suggestion
package cosmos.crypto.hd.v1;
```

amaurym commented on crypto/keyring/legacy_info.go:
```diff
@@ -1,50 +1,51 @@
 package keyring
```
how about doing like in module, and creating a `crypto/keyring/migrations` package. Then all migration logic (including this legacyInfo) can go in that package.

cyberbono3 replied:
what do you think, @robert-zaremba ?

cyberbono3 replied:
> how about doing like in module, and creating a `crypto/keyring/migrations` package. Then all migration logic (including this legacyInfo) can go in that package.

What do you wanna to put inside `crypto/keyring/migrations` package? `record.go`, `record_test.go` and `legacy_info.go` ? It is worth noting that `Migrate` and `MigrateAll()` methods are located inside `keyring` package.

cyberbono3 replied:
```
func (ks keystore) MigrateAll() (bool, error) {
    return migrations.MigrateAll(ks.db, ks.cdc)
}
```

I started moving legacy_info.go and migration logic (`MigrateAll` and `Migrate` funcs from keyring.go) into `crypto/keyring/migrations` package. if i dont move `record.go` into `migrations` package, it introduces `import cycle dep  issue` since I require access to `Record` struct inside `migrations` package . May I move 'record.go` into `migrations` package as well? It would be great to move `types.pb.go` and `types.go` to `migrations` package as well.

robert-zaremba replied:
I like the idea of `migrations` package. But I think it will be challenging without breaking `keyring` package in sub-packages.
@cyberbono3  - migrations package shouldn't contain core types. It should only contain migration logic and legacy types.

So, if it will be too much of efforts, then I think it's not really worth.

cyberbono3 replied:
It is challenging due to cycle dep issue described above. From my perspective, it is not worth it in this case.

amaurym replied:
Sounds good

amaurym commented on client/keys/add.go:
```diff
@@ -118,8 +118,7 @@ func runAddCmd(ctx client.Context, cmd *cobra.Command, args []string, inBuf *buf
 	if dryRun, _ := cmd.Flags().GetBool(flags.FlagDryRun); dryRun {
-		// use in memory keybase
-		kb = keyring.NewInMemory()
+		kb = keyring.NewInMemory(ctx.Codec)
```
let's keep the old comment

```suggestion
		// use in memory keybase
		kb = keyring.NewInMemory(ctx.Codec)
```

amaurym commented on proto/cosmos/crypto/hd/hd.proto:
```diff
@@ -0,0 +1,18 @@
+	uint32 coin_type = 2;
+	uint32 account = 3;
+	bool   change = 4;
+	uint32 address_index = 5;
```
could we add comments to what each field means?

amaurym commented on crypto/keyring/signing_algorithms_test.go:
```diff
@@ -51,25 +51,25 @@ func TestAltSigningAlgoList_Contains(t *testing.T) {
 }
 
-type notSupportedAlgo struct {
+type NotSupportedAlgo struct {
```
why do we need this change?

cyberbono3 replied:
It is used in this [test](https://github.com/cosmos/cosmos-sdk/blob/cyberbono3/keyring-migration/crypto/keyring/keyring_ledger_test.go#L105)

amaurym commented on crypto/keyring/record.go:
```diff
@@ -0,0 +1,125 @@
+	return ""
+}
+
+// TODO consider to remove it
```
Should we address this TODO?

amaurym commented on simapp/simd/cmd/root.go:
```diff
@@ -46,7 +47,8 @@ func NewRootCmd() (*cobra.Command, params.EncodingConfig) {
 		WithHomeDir(simapp.DefaultNodeHome).
-		WithViper("") // In simapp, we don't use any prefix for env variables.
+		WithViper(""). // In simapp, we don't use any prefix for env variables.
+		WithCodec(encodingConfig.Codec)
```
there's already a WithCodec on L43

amaurym commented on simapp/simd/cmd/testnet.go:
```diff
@@ -253,7 +254,9 @@ func initTestnetFiles(
-		kb, err := keyring.New(sdk.KeyringServiceName(), args.keyringBackend, nodeDir, inBuf)
+		cdc := simapp.MakeTestEncodingConfig().Codec
+
+		kb, err := keyring.New(sdk.KeyringServiceName(), args.keyringBackend, nodeDir, inBuf, cdc)
```
```suggestion
		kb, err := keyring.New(sdk.KeyringServiceName(), args.keyringBackend, nodeDir, inBuf, clientCtx.Codec)
```

amaurym replied:
same here, let's not create new codecs (except in tests)

amaurym commented on testutil/network/network.go:
```diff
@@ -338,7 +338,8 @@ func New(l Logger, baseDir string, cfg Config) (*Network, error) {
 		valPubKeys[i] = pubKey
 
-		kb, err := keyring.New(sdk.KeyringServiceName(), keyring.BackendTest, clientDir, buf, cfg.KeyringOptions...)
+		encCfg := simapp.MakeTestEncodingConfig()
+		kb, err := keyring.New(sdk.KeyringServiceName(), keyring.BackendTest, clientDir, buf, encCfg.Codec, cfg.KeyringOptions...)
```
```suggestion
		kb, err := keyring.New(sdk.KeyringServiceName(), keyring.BackendTest, clientDir, buf, cfg.Codec, cfg.KeyringOptions...)
```

amaurym replied:
still relevant, please don't create a new Codec here

amaurym commented on types/module/module_test.go:
```diff
@@ -5,9 +5,10 @@ import (
-	"github.com/cosmos/cosmos-sdk/codec/types"
 	"github.com/gorilla/mux"
 
+	"github.com/cosmos/cosmos-sdk/codec/types"
```
this import should go with other cosmos ones

amaurym commented on x/distribution/keeper/msg_server.go:
```diff
@@ -3,7 +3,7 @@ package keeper
 	"context"
 
-	"github.com/armon/go-metrics"
+	metrics "github.com/armon/go-metrics"
```
why is this needed?

amaurym commented on x/gov/client/testutil/deposits.go:
```diff
@@ -4,13 +4,14 @@ import (
 	"fmt"
 	"time"
 
+	"github.com/stretchr/testify/suite"
+	tmcli "github.com/tendermint/tendermint/libs/cli"
```
while these changes are correct, it makes revieweing hard (82 files changed). If possible, in the future, let's avoid making side-changes in a big PR, prefer to have 1 PR that's only focused on fixing lints

cyberbono3 replied:
I do not get what you mean here. Could you clarify, please?

amaurym commented on x/gov/keeper/msg_server.go:
```diff
@@ -5,7 +5,7 @@ import (
 	"fmt"
 	"strconv"
 
-	"github.com/armon/go-metrics"
+	metrics "github.com/armon/go-metrics"
```
same here

amaurym commented on x/genutil/utils.go:
```diff
@@ -6,7 +6,7 @@ import (
 	"time"
 
-	"github.com/cosmos/go-bip39"
+	bip39 "github.com/cosmos/go-bip39"
```
```suggestion
	"github.com/cosmos/go-bip39"
```

amaurym commented on x/genutil/client/cli/init.go:
```diff
@@ -16,6 +15,8 @@ import (
 	tmrand "github.com/tendermint/tendermint/libs/rand"
 	"github.com/tendermint/tendermint/types"
 
+	bip39 "github.com/cosmos/go-bip39"
```
this should be reverted

amaurym reviewed (commented):
some additional comments (still didn't review the whole PR).

cyberbono3 commented on crypto/keyring/record.go:
```diff
@@ -62,9 +79,19 @@ func (k Record) GetAddress() (types.AccAddress, error) {
 }
 
-// TODO consider to remove it
 func (k Record) GetType() KeyType {
```
Since `Record.GetType()` method is used in [readTxCommandFlags](https://github.com/cosmos/cosmos-sdk/blob/cyberbono3/keyring-migration/client/cmd.go#L243) and in [GetFromFields](https://github.com/cosmos/cosmos-sdk/blob/cyberbono3/keyring-migration/client/context.go#L350) I decided to modify this method and leave as it is. What do you think, @amaurym ?

cyberbono3 replied:
After discussion with @amaurym, we decided to leave `Record.GetType()`

amaurym commented on crypto/keyring/keyring.go:
```diff
@@ -217,95 +242,81 @@ func newKeystore(kr keyring.Keyring, opts ...Option) keystore {
 	}
 
-	return crypto.ArmorPubKeyBytes(legacy.Cdc.MustMarshal(bz.GetPubKey()), string(bz.GetAlgo())), nil
+	return crypto.ArmorPubKeyBytes(legacy.Cdc.MustMarshal(key), key.Type()), nil
```
When exporting pubkeys, we should also use ks.cdc (proto)

cyberbono3 replied:
I believe, this is good way to go:
```
	bz, err := ks.cdc.MarshalInterface(key)
	if err != nil {
		return "", err
	}

	return crypto.ArmorPubKeyBytes(bz, key.Type()), nil
}

```

amaurym commented on crypto/keyring/keyring.go:
```diff
@@ -217,95 +242,81 @@ func newKeystore(kr keyring.Keyring, opts ...Option) keystore {
+	return ks.ExportPubKeyArmor(k.Name)
 }
 
+// we use ExportPrivateKeyFromLegacyInfo(info LegacyInfo) (cryptotypes.PrivKey, error) { for LegacyInfo
```
Godocs should start with `ExportPrivKeyArmor...`

amaurym commented on crypto/keyring/keyring.go:
```diff
@@ -217,95 +242,81 @@ func newKeystore(kr keyring.Keyring, opts ...Option) keystore {
+// we use ExportPrivateKeyFromLegacyInfo(info LegacyInfo) (cryptotypes.PrivKey, error) { for LegacyInfo
 func (ks keystore) ExportPrivKeyArmor(uid, encryptPassphrase string) (armor string, err error) {
-	priv, err := ks.ExportPrivateKeyObject(uid)
+	fmt.Println("ExportPrivKeyArmor start")
```
```suggestion
```

amaurym commented on crypto/keyring/keyring.go:
```diff
@@ -217,95 +242,81 @@ func newKeystore(kr keyring.Keyring, opts ...Option) keystore {
 	if err != nil {
 		return "", err
 	}
+	fmt.Println("ExportPrivKeyArmor done")
```
```suggestion
```

amaurym commented on crypto/keyring/keyring.go:
```diff
@@ -217,95 +242,81 @@ func newKeystore(kr keyring.Keyring, opts ...Option) keystore {
 // ExportPrivateKeyObject exports an armored private key object.
-func (ks keystore) ExportPrivateKeyObject(uid string) (types.PrivKey, error) {
-	info, err := ks.Key(uid)
+func (ks keystore) ExportPrivateKeyObject(uid string) (*Record, types.PrivKey, error) {
```
Is there a reason why we changed the function signature? I would keep the old one

cyberbono3 replied:
yeah it does make sense!

amaurym commented on crypto/keyring/keyring.go:
```diff
@@ -337,38 +348,43 @@ func (ks keystore) ImportPubKey(uid string, armor string) error {
-func (ks keystore) ImportInfo(oldInfo Info) error {
-	if _, err := ks.Key(oldInfo.GetName()); err == nil {
-		return fmt.Errorf("cannot overwrite key: %s", oldInfo.GetName())
+// TODO do i need it or not?
```
let's remove?

amaurym commented on proto/cosmos/crypto/hd/v1/hd.proto:
```diff
@@ -0,0 +1,22 @@
+	uint32 account = 3;
+	// change is a constant used for public derivation. Constant 0 is used for external chain and constant 1 for internal chain.
+	bool   change = 4;
+	// address_index is used as child index in BIP32 derivation
```
nice comments!

amaurym reviewed (changes requested):
Gave a thorough review of about half of the PR, there's still some work. Most notably, I found some instances of legacy Cdc used; that should only be used inside migrations.

@cyberbono3 Some suggestions to make reviews easier:
- Make sure all TODOs that you left in the code are addressed.
- Try to aim for smallest code diff. For example, I'm not sure why some `fmt.Sprintf("--%s=%s", flags.FlagKeyringBackend, keyring.BackendTest)` were removed from tests

cyberbono3 commented on client/context.go:
```diff
@@ -329,27 +329,32 @@ func GetFromFields(kr keyring.Keyring, from string, genOnly bool) (sdk.AccAddres
 func NewKeyringFromBackend(ctx Context, backend string) (keyring.Keyring, error) {
 	if ctx.GenerateOnly || ctx.Simulate {
-		return keyring.New(sdk.KeyringServiceName(), keyring.BackendMemory, ctx.KeyringDir, ctx.Input, ctx.KeyringOptions...)
+		return keyring.New(sdk.KeyringServiceName(), keyring.BackendMemory, ctx.KeyringDir, ctx.Input, ctx.Codec, ctx.KeyringOptions...)
```
Should i add them to unreleased section of changelog?

amaurym replied:
yes!

cyberbono3 commented on client/keys/delete_test.go:
```diff
@@ -51,13 +53,12 @@ func Test_runDeleteCmd(t *testing.T) {
 	cmd.SetArgs([]string{
 		fakeKeyName1,
 		fmt.Sprintf("--%s=%s", flags.FlagHome, kbHome),
-		fmt.Sprintf("--%s=%s", flags.FlagKeyringBackend, keyring.BackendTest),
```
I have deleted them to make this test pass. As you can see [here](https://github.com/cosmos/cosmos-sdk/pull/9695/commits/193d8107d6983e8043e36d2fac91440ca8df7721) I added `cdc` to `clientCtx` and  added`flags.FlagKeyringBackend` to `cmd.SetArgs` back to make it work.

cyberbono3 commented on client/keys/add_ledger_test.go:
```diff
@@ -160,12 +169,14 @@ func Test_runAddCmdLedgerDryRun(t *testing.T) {
 			kbHome := t.TempDir()
 			mockIn := testutil.ApplyMockIODiscardOutErr(cmd)
-			kb, err := keyring.New(sdk.KeyringServiceName(), keyring.BackendTest, kbHome, mockIn)
+			cdc := simapp.MakeTestEncodingConfig().Codec
```
it makes sense!

cyberbono3 commented on client/keys/utils.go:
```diff
@@ -51,8 +37,8 @@ func printKeyInfo(w io.Writer, keyInfo cryptokeyring.Info, bechKeyOut bechKeyOut
 
-func printInfos(w io.Writer, infos []cryptokeyring.Info, output string) {
-	kos, err := cryptokeyring.MkAccKeysOutput(infos)
+func printKeyringRecords(w io.Writer, records []*cryptokeyring.Record, output string) {
```
should I use ` codec.ProtoMarshalJSON` for that?

cyberbono3 commented on crypto/keyring/keyring.go:
```diff
@@ -122,14 +127,33 @@ type Importer interface {
 	// It is used to migrate multisig, ledger, and public key Info structure.
-	ImportInfo(oldInfo Info) error
+	ImportInfo(oldInfo LegacyInfo) error
+}
```
Sure!

cyberbono3 commented on crypto/keyring/keyring.go:
```diff
@@ -122,14 +127,33 @@ type Importer interface {
+type Marshaler interface {
+	ProtoMarshalRecord(k *Record) ([]byte, error)
+	ProtoUnmarshalRecord(bz []byte) (*Record, error)
 }
```
Renaming the package might cause cycle dep issue. I am not sure if it is a good good solution. Another possible solution might be to make `keystore.db` field public, so the `keyring_test` package can access it.
```

cyberbono3 replied:
I encountered cycle dep issue due to importing `simapp`

Running tool: /usr/local/go/bin/go test -timeout 120s -run ^TestNewKeyring$ github.com/cosmos/cosmos-sdk/crypto/keyring -tags='ledger test_ledger_mock norace experimental'
```
# github.com/cosmos/cosmos-sdk/crypto/keyring
package github.com/cosmos/cosmos-sdk/crypto/keyring (test)
	imports github.com/cosmos/cosmos-sdk/simapp
	imports github.com/cosmos/cosmos-sdk/baseapp
	imports github.com/cosmos/cosmos-sdk/client
	imports github.com/cosmos/cosmos-sdk/client/flags
	imports github.com/cosmos/cosmos-sdk/crypto/keyring: import cycle not allowed in test
FAIL	github.com/cosmos/cosmos-sdk/crypto/keyring [setup failed]
FAIL
```

cyberbono3 replied:
Since I import `simapp` only to get codec, i can implement `getCodec()` function to avoid this import and cycle dep issue.
```
func getCodec() codec.Codec {
  registry := codectypes.NewInterfaceRegistry()
 cryptocodec.RegisterInterfaces(registry)
 return codec.NewProtoCodec(registry)
}

```

What do you think, @amaury?

amaurym replied:
Yes, that sounds like a good idea!

cyberbono3 commented:
`TODO`  scan all code in keyring.go and replace legacy codec to proto codec

cyberbono3 commented on crypto/keyring/keyring.go:
```diff
@@ -122,11 +125,13 @@ type Importer interface {
+
+type Setter interface {
+	// Set item in keyring.Item
+	SetItem(item keyring.Item) error
```
Since `SetItem` is used  in tests, `keyring` and `keys` package, i decided to declare `Setter` interface and make `SetItem` public.

cyberbono3 commented on crypto/keyring/legacy_info.go:
```diff
@@ -187,68 +193,68 @@ type multisigPubKeyInfo struct {
 
 // multiInfo is the public information about a multisig key
-type multiInfo struct {
+type LegacyMultiInfo struct {
```
Just wondering since `LegacyMultiInfo` is a public struct, should `legacyInfo` be public interface which is implemented by `LegacyMultiInfo`?

cyberbono3 commented:
> * ExportPrivateKeyObject

Thanks for your review..
I have addressed all TODOs in `keyring` package. i have investigated minor diffs and added changelog entries for API breaking changes. 

There are some linter issues that do not hold true. 
 1. Error: `newLegacyLocalInfo` is unused (deadcode)
 2. Error: `newLegacyLedgerInfo` is unused (deadcode)
 3. Error: `newLegacyOfflineInfo` is unused (deadcode)
I have double checked. All these 3 functions are indeed used in a code.

Another linter issue sounds questionable to me:
1. `exported func NewLegacyMultiInfo returns unexported type github.com/cosmos/cosmos-sdk/crypto/keyring.legacyInfo, which can be annoying to use (golint)`. 
Do you think, it is a good practice to return a private data type in this case? or I should make `legacyInfo` public?

cyberbono3 commented on crypto/codec/proto.go:
```diff
@@ -16,5 +16,10 @@ func RegisterInterfaces(registry codectypes.InterfaceRegistry) {
+	var priv *cryptotypes.PrivKey
+	registry.RegisterInterface("cosmos.crypto.PrivKey", priv)
+	registry.RegisterImplementations(priv, &secp256k1.PrivKey{})
+	registry.RegisterImplementations(priv, &ed25519.PrivKey{}) //nolint
```
I added `nolint` to resolve this issue. Does it sounds good?
```
SA1019: ed25519.PrivKey is deprecated: PrivKey defines a ed25519 private key. NOTE: ed25519 keys must not be used in SDK apps except in a tendermint validator context. 
```

cyberbono3 commented on crypto/keyring/legacy_info.go:
```diff
@@ -44,7 +44,8 @@ type legacyLocalInfo struct {
 
-func newLegacyLocalInfo(name string, pub cryptotypes.PubKey, privArmor string, algo hd.PubKeyType) legacyInfo {
+//nolint
+func newLegacyLocalInfo(name string, pub cryptotypes.PubKey, privArmor string, algo hd.PubKeyType) legacyInfo { 
```
Added `nolint` cause `newLegacyLocalInfo` is called only in test.

cyberbono3 replied:
move it to test file

cyberbono3 commented on crypto/keyring/keyring.go:
```diff
@@ -494,7 +494,7 @@ func (ks keystore) List() ([]*Record, error) {
 	}
 
-	var res []*Record
+	var res []*Record //nolint
```
linter suggest preallocate`res` like `res := make([]*Record, len(keys))` that will cause some refactoring and fixing tests, do you think it is a good idea?

cyberbono3 replied:
Done

cyberbono3 commented:
linter outputs issues but does not say which line :
```
 Running [/home/runner/golangci-lint-1.39.0-linux-amd64/golangci-lint run --out-format=github-actions --timeout 10m] in [] ...
  Error: directive `//nolint` is unused (nolintlint)
  Error: directive `//nolint` is unused (nolintlint)
  
  Error: issues found
```
It looks weird.
I run `golangci-lint run` and it does not yield any issues.

cyberbono3 commented on crypto/keyring/legacy_info.go:
```diff
@@ -12,8 +12,8 @@ import (
-// Info is the publicly exposed information about a keypair
-type legacyInfo interface {
+// LegacyInfo is the publicly exposed information about a keypair
+type LegacyInfo interface {
```
I made LegacyInfo public to resolve linter issue `exported func NewLegacyMultiInfo returns unexported type github.com/cosmos/cosmos-sdk/crypto/keyring.legacyInfo, which can be annoying to use`

robert-zaremba replied:
Do we need to export `NewLegacyMultiInfo`? Maybe we can remove it?

cyberbono3 replied:
we should export `NewLegacyMultiInfo` cause it is used in `client/keys` package.

cyberbono3 commented on client/context.go:
```diff
@@ -329,27 +329,32 @@ func GetFromFields(kr keyring.Keyring, from string, genOnly bool) (sdk.AccAddres
 	}
 
-	var info keyring.Info
+	k := new(keyring.Record) //nolint
```
I put `nolint` to resolve liner issue `k is not used in code`

amaurym replied:
OK, this is weird, because `k` is obviously used in  L354

cyberbono3 commented on crypto/keyring/keyring_test.go:
```diff
@@ -29,69 +33,118 @@ func init() {
 	crypto.BcryptSecurityParameter = 1
 }
 
+func getCodec() codec.Codec {
```
I declared `getCodec()` to fix cycle dep issue caused by importing `simapp`  that is used to call `simapp.MakeTestEncodingConfig().Codec`

robert-zaremba replied:
good, in fact we have a task to reduce `MakeTestEncodingConfig` usage.

cyberbono3 commented on crypto/keyring/keyring_test.go:
```diff
@@ -788,41 +954,65 @@ func ExampleNew() {
 	// signed by Bob
 }
 
+func getIndices(list []*Record, uid1, uid2, uid3 string) (i, j, k int) {
```
`getIndices()` is used to fetch indices that are used in `TestAltKeyring_List`

amaurym replied:
This indices manipulation is really confusing IMO. Is there some way to return in some deterministic order (like alphabetical like before)?

cyberbono3 replied:
I came up with an idea to use  `func getIndices(list []*Record, uids ...string) map[string]int ` to return a `hashmap[uid] = index` .  You can find more details [here](https://github.com/cosmos/cosmos-sdk/blob/cyberbono3/keyring-migration/crypto/keyring/keyring_test.go#L76). Does it sounds good to you?
```


```

amaurym commented on client/keys/add_ledger_test.go:
```diff
@@ -107,9 +112,10 @@ func Test_runAddCmdLedger(t *testing.T) {
 
-	// Now check that it has been stored properly
-	kb, err := keyring.New(sdk.KeyringServiceName(), keyring.BackendTest, kbHome, mockIn)
+	kb, err := keyring.New(sdk.KeyringServiceName(), keyring.BackendTest, kbHome, mockIn, encCfg.Codec)
```
```suggestion
	// Now check that it has been stored properly
	kb, err := keyring.New(sdk.KeyringServiceName(), keyring.BackendTest, kbHome, mockIn, encCfg.Codec)
```
let's keep the old comment

amaurym commented on client/keys/migrate.go:
```diff
@@ -1,39 +1,29 @@
-is correct. Otherwise, the command will exit and migration must be repeated.
+		Use:   "migrate",
+		Short: "Migrate keys from amino to proto serialization format",
+		Long: `Migrate key information from LegacyInfo(amino) to record(proto).
```
```suggestion
		Long: `Migrate key information from LegacyInfo(amino) to Record(proto).
```

amaurym commented on crypto/keyring/record.go:
```diff
@@ -0,0 +1,132 @@
+
+var ErrPrivKeyExtr = errors.New("private key extraction works only for Local")
+
+func NewRecord(name string, pk cryptotypes.PubKey, item isRecord_Item) (*Record, error) {
```
There are a lof of `New*` functions here, and it's a bit confusing.

I propose only four public functions in this file:
```go
func NewLocalRecord(name, pub cryptotypes.PubKey, priv cryptotypes.PrivKey) (*Record, error)
func NewLedgerRecord(name, pub cryptotypes.PubKey, path hd.BIP44Params) (*Record, error)
func NewMultiRecord(name, pub cryptotypes.PubKey) (*Record, error)
func NewOfflineRecord(name, pub cryptotypes.PubKey) (*Record, error)
```

and all other functions (if needed) are private

amaurym commented on client/keys/show.go:
```diff
@@ -53,36 +51,43 @@ consisting of all the keys provided by name and multisig threshold.`,
 
 		multikey := multisig.NewLegacyAminoPubKey(multisigThreshold, pks)
-		info, err = keyring.NewMultiInfo(defaultMultiSigKeyName, multikey)
+		emptyRecord := keyring.NewMultiRecord()
+		emptyRecordItem := keyring.NewMultiRecordItem(emptyRecord)
+
+		k, err = keyring.NewRecord(k.Name, multikey, emptyRecordItem)
```
I see your're doing the `emptyRecord` hack multiple times, it seems weird to me. Does my comment above work? in this case we would do:
```suggestion
		k, err = keyring.NewMultiRecord(k.Name, multikey)
```

cyberbono3 replied:
Yeah. Comment above works.

amaurym commented on client/keys/show_test.go:
```diff
@@ -24,7 +25,7 @@ func Test_multiSigKey_Properties(t *testing.T) {
 		[]cryptotypes.PubKey{tmpKey1.PubKey()},
 	)
-	tmp, err := keyring.NewMultiInfo("myMultisig", pk)
+	tmp, err := keyring.NewLegacyMultiInfo("myMultisig", pk)
```
Why are we creating a legacy info here? Does this test run migration?

cyberbono3 replied:
Done

amaurym commented on client/keys/utils.go:
```diff
@@ -51,8 +37,9 @@ func printKeyInfo(w io.Writer, keyInfo cryptokeyring.Info, bechKeyOut bechKeyOut
 
-func printInfos(w io.Writer, infos []cryptokeyring.Info, output string) {
-	kos, err := cryptokeyring.MkAccKeysOutput(infos)
+// put first argument cdc
```
```suggestion
```

amaurym commented on client/keys/utils.go:
```diff
@@ -62,6 +49,7 @@ func printInfos(w io.Writer, infos []cryptokeyring.Info, output string) {
 		printTextInfos(w, kos)
 
 	case OutputFormatJSON:
+		// TODO https://github.com/cosmos/cosmos-sdk/issues/8046
```
```suggestion
		// TODO https://github.com/cosmos/cosmos-sdk/issues/8046
		// Replace AminoCdc with Proto JSON
```

amaurym commented on cosmovisor/upgrade.go:
```diff
@@ -11,8 +11,8 @@ import (
 
-	"github.com/hashicorp/go-getter"
 	"github.com/otiai10/copy"
+	"github.com/hashicorp/go-getter"
```
This doesn't seem go-formatted

amaurym commented on crypto/keyring/keyring.go:
```diff
@@ -122,11 +124,8 @@ type Importer interface {
-	// ImportInfo import a keyring.Info into the current keyring.
-	// It is used to migrate multisig, ledger, and public key Info structure.
-	ImportInfo(oldInfo Info) error
+type Migrator interface {
```
Let's add godocs for public interfaces

amaurym commented on crypto/keyring/keyring_test.go:
```diff
@@ -7,10 +7,14 @@ import (
 
 	"github.com/cosmos/cosmos-sdk/crypto"
+	"github.com/cosmos/go-bip39"
+	"github.com/stretchr/testify/require"
```
the imports were correct before

amaurym commented on crypto/keyring/keyring_test.go:
```diff
@@ -29,69 +33,118 @@ func init() {
-	require.Equal(t, n2, keyS[0].GetName())
-	require.Equal(t, n1, keyS[1].GetName())
-	require.Equal(t, i2.GetPubKey(), keyS[0].GetPubKey())
+	require.True(t, requireEqualNames(keyS, n1, n2))
+
+	key1, err := k2.GetPubKey()
+	require.NoError(t, err)
+	require.NotNil(t, key1)
+	i := getN2Index(keyS, n2)
```
Are the key names not in alphabetical order anymore? Why do we need requireEqualNames and getN2Index?

cyberbono3 replied:
I tried to follow @amaurym recommendation and remove both `requireEqualNames` and `getIndices()`, but unfortunately it did not work out. The reason for that is the indices in `res`  inside `keyring.List()` are being assigned non deterministically while running `TestKeyManagementKeyRing` multiple times.

amaurym commented on crypto/keyring/legacy_info.go:
```diff
@@ -1,179 +1,159 @@
 
-// Info is the publicly exposed information about a keypair
-type Info interface {
+// LegacyInfo is the publicly exposed information about a keypair
```
```suggestion
// Deprecated: LegacyInfo is the publicly exposed information about a keypair
```

amaurym commented on client/keys/migrate_test.go:
```diff
@@ -3,38 +3,130 @@ package keys
+			priv.PubKey(),
+		},
+	)
+	LegacyMultiInfo, err := keyring.NewLegacyMultiInfo(n1, multi)
```
```suggestion
	legacyMultiInfo, err := keyring.NewLegacyMultiInfo(n1, multi)
```

amaurym commented on x/staking/client/testutil/grpc.go:
```diff
@@ -358,10 +358,11 @@ func (s *IntegrationTestSuite) TestGRPCQueryDelegatorDelegations() {
 
-	// Create new account in the keyring for address without delegations.
-	info, _, err := val.ClientCtx.Keyring.NewMnemonic("test", keyring.English, sdk.FullFundraiserPath, keyring.DefaultBIP39Passphrase, hd.Secp256k1)
+	// Create new account in the keyring.
```
```suggestion
	// Create new account in the keyring for address without delegations.
```

Let's keep the same comment as before

amaurym commented on x/authz/client/testutil/tx.go:
```diff
@@ -61,6 +61,7 @@ func (s *IntegrationTestSuite) SetupSuite() {
 	// Create new account in the keyring.
 	s.grantee[1] = s.createAccount("grantee2")
 	// Send some funds to the new account.
+	fmt.Println("s.msgSendExec(s.grantee[1])")
```
```suggestion
```

amaurym commented on x/authz/client/testutil/tx.go:
```diff
@@ -84,9 +85,11 @@ func (s *IntegrationTestSuite) SetupSuite() {
 	s.Require().NoError(err)
-	return sdk.AccAddress(info.GetPubKey().Address())
+
+	addr, _ := k.GetAddress()
```
let's check the error here

amaurym commented on simapp/simd/cmd/root.go:
```diff
@@ -6,13 +6,14 @@ import (
 	dbm "github.com/tendermint/tm-db"
 
+	serverconfig "github.com/cosmos/cosmos-sdk/server/config"
+
```
```suggestion
```

amaurym commented on simapp/simd/cmd/genaccounts.go:
```diff
@@ -53,18 +54,23 @@ contain valid denominations. Accounts may optionally be supplied with vesting pa
 				inBuf := bufio.NewReader(cmd.InOrStdin())
 				keyringBackend, _ := cmd.Flags().GetString(flags.FlagKeyringBackend)
 
+				cdc := simapp.MakeTestEncodingConfig().Codec
```
same here, don't create new Codec

amaurym commented on server/rosetta/converter.go:
```diff
@@ -264,6 +264,7 @@ func (c converter) Ops(status string, msg sdk.Msg) ([]*rosettatypes.Operation, e
 	return ops, nil
 }
 
+//nolint
```
why this change?

amaurym commented on server/init.go:
```diff
@@ -5,18 +5,24 @@ import (
 func GenerateCoinKey(algo keyring.SignatureAlgo) (sdk.AccAddress, string, error) {
 	// generate a private key, with recovery phrase
-	info, secret, err := keyring.NewInMemory().NewMnemonic("name", keyring.English, sdk.FullFundraiserPath, keyring.DefaultBIP39Passphrase, algo)
+	encCfg := simapp.MakeTestEncodingConfig()
```
I don't think it's a good idea to create a new Codec here. Maybe let's add a `cdc Codec` argument to this function

cyberbono3 replied:
Sounds good

amaurym commented on proto/cosmos/crypto/keyring/v1/types.proto:
```diff
@@ -0,0 +1,45 @@
+option go_package = "github.com/cosmos/cosmos-sdk/crypto/keyring";
+option (gogoproto.goproto_getters_all) = false;
+
+// Record is used for keyring migration from LegacyInfo to proto.
```
```suggestion
// Record is used for representing a key in the keyring.
```

amaurym commented on proto/cosmos/crypto/keyring/v1/types.proto:
```diff
@@ -0,0 +1,45 @@
+
+// Record is used for keyring migration from LegacyInfo to proto.
+message Record {
+  string name = 1;
+  google.protobuf.Any pub_key = 2;
```
Let's add some comments for these 2 fields

amaurym reviewed (commented):
I reviewed `60 / 85` files, and left some comments. An anti-pattern I'm seeing is to create `simapp.MakeTestEncConfig()`, which should be avoided except in tests.

cyberbono3 commented on CHANGELOG.md:
```diff
@@ -38,15 +38,33 @@ Ref: https://keepachangelog.com/en/1.0.0/
+* [\#9695](https://github.com/cosmos/cosmos-sdk/pull/9695) `MkAccKeysOutput` take slice of `*keyring.Record` as first argument instead of slice of`Info`.
+* [\#9695](https://github.com/cosmos/cosmos-sdk/pull/9695)`MkConsKeyOutput`, ` MkValKeyOutput` and `MkAccKeyOutput` take `*keyring.Record` as first argument instead of `Info`.
+* [\#9695](https://github.com/cosmos/cosmos-sdk/pull/9695)`SignWithLedger` takes `*keyring.Record` as first argument instead of `Info`.
+* [\#9695](https://github.com/cosmos/cosmos-sdk/pull/9695)`NewMnemonic` now returns `keyring.Record` instead of `Info`.
```
Done!

cyberbono3 commented on CHANGELOG.md:
```diff
@@ -38,15 +38,33 @@ Ref: https://keepachangelog.com/en/1.0.0/
 
 ### API Breaking Changes
-
+* [\#9695](https://github.com/cosmos/cosmos-sdk/pull/9695) Remove `algo` argument from `writeLocalKey`, `writeLedgerKey` and `writeOfflineKey`.
```
Yes!

robert-zaremba commented on crypto/keyring/keyring.go:
```diff
@@ -318,57 +308,50 @@ func (ks keystore) ImportPubKey(uid string, armor string) error {
-	case offlineInfo, multiInfo:
-		return nil, info.GetPubKey(), errors.New("cannot sign with offline keys")
+		return nil, pub, errors.New("cannot sign with offline keys")
 	}
```
```suggestion
	}
	if err != nil {
		return nil, nil, err
	}
```

robert-zaremba commented on crypto/keyring/keyring.go:
```diff
@@ -318,57 +308,50 @@ func (ks keystore) ImportPubKey(uid string, armor string) error {
+	switch {
+	case k.GetLocal() != nil:
+		priv, err = extractPrivKeyFromLocal(k.GetLocal())
+		if err != nil {
+			return nil, nil, err
 		}
 
```
```suggestion
```

robert-zaremba replied:
we can move err check after the switch.

robert-zaremba commented on crypto/keyring/keyring.go:
```diff
@@ -582,29 +575,31 @@ func (ks keystore) NewAccount(name string, mnemonic string, bip39Passphrase stri
-func (ks keystore) key(infoKey string) (Info, error) {
-	bs, err := ks.db.Get(infoKey)
+func (ks keystore) Key(uid string) (*Record, error) {
+	if _, err := ks.migrate(uid); err != nil {
```
`migrate(uid)` can return a `*Record` - so we don't need the code below to re query the key.

robert-zaremba replied:
`migrate(uid)` can return a `*Record` or `Item` - so we don't need the code below to re query the key.

robert-zaremba commented on crypto/keyring/keyring.go:
```diff
@@ -842,26 +837,166 @@ func (ks keystore) existsInDb(info Info) (bool, error) {
+}
+
+// for one key
+func (ks keystore) migrate(key string) (bool, error) {
```
let's return `Item, bool, error`

cyberbono3 replied:
from my perspective, it is more efficient to return `Record,bool,error`

robert-zaremba commented on crypto/keyring/keyring.go:
```diff
@@ -842,26 +837,166 @@ func (ks keystore) existsInDb(info Info) (bool, error) {
+
+// for one key
+func (ks keystore) migrate(key string) (bool, error) {
+	var migrated bool
```
this variable is never assigned. Let's remove it and be explicit (returning true / false) in all return statements below.

robert-zaremba commented on crypto/keyring/keyring.go:
```diff
@@ -775,65 +766,69 @@ func newRealPrompt(dir string, buf io.Reader) func(string) (string, error) {
-func (ks keystore) writeInfo(info Info) error {
-	key := infoKeyBz(info.GetName())
-	serializedInfo := marshalInfo(info)
+func (ks keystore) writeRecord(k *Record) error {
```
let's add a go doc about the proper conditions for using this functions (eg: that the key must not exists).

robert-zaremba commented on crypto/keyring/keyring.go:
```diff
@@ -842,26 +837,166 @@ func (ks keystore) existsInDb(info Info) (bool, error) {
+	}
+
+	if len(keys) == 0 {
+		fmt.Println("no keys available for migration")
```
I  think we should remove this debug print.

robert-zaremba commented on crypto/keyring/keyring.go:
```diff
@@ -842,26 +837,166 @@ func (ks keystore) existsInDb(info Info) (bool, error) {
+}
+
+func (ks keystore) MigrateAll() (bool, error) {
+	var migrated bool
```
please move this line just above the for loop, and in between in return statements use explicit `false`.

robert-zaremba commented on crypto/keyring/keyring.go:
```diff
@@ -842,26 +837,166 @@ func (ks keystore) existsInDb(info Info) (bool, error) {
+			continue
+		}
+
+		migrated, err = ks.migrate(key)
+		if err != nil {
+			fmt.Printf("migrate err: %q", err)
+			continue
+		}
```
```suggestion
		if migrated2, err = ks.migrate(key); err != nil {
			fmt.Printf("migrate err: %q", err)
			continue
		} else if migrated2 {
			migrated = true
		}
```

robert-zaremba replied:
The function should return true if at least one migration was done.

cyberbono3 replied:
sounds good

robert-zaremba commented on crypto/keyring/keyring.go:
```diff
@@ -842,26 +837,166 @@ func (ks keystore) existsInDb(info Info) (bool, error) {
+		k, err := NewLocalRecord(name, priv, pk)
+		if err != nil {
+			return nil, err
+		}
```
how about returning directly:
```
return NewLocalRecord(name, priv, pk)
```
Same for the cases below.

cyberbono3 replied:
Done!

robert-zaremba commented on crypto/keyring/keyring_test.go:
```diff
@@ -29,69 +32,125 @@ func init() {
+	require.Equal(t, "foo", k.Name)
 }
 
+func requireEqualNames(records []*Record, n1, n2 string) bool {
```
```suggestion
func requirelNameOneOf(records []*Record, n1, n2 string) bool {
```

robert-zaremba commented on crypto/keyring/migration_test.go:
```diff
@@ -0,0 +1,357 @@
+
+	migrated, err := ks.migrate(n1)
+	require.True(migrated)
+	require.NoError(err)
```
same in the test cases below

robert-zaremba replied:
for testing the `ks.Key` we can use one common helper function in all the tests below.

cyberbono3 replied:
`func (ks keystore) migrate(key string) (*Record, bool, error)` 
if i wanna check the status of migration I have to check 2nd return bool value. From my perspective, usage of  `ks.Key` is not appropriate in this case. 
```
_, migrated, err := s.ks.migrate(n1)
s.Require().True(migrated)
s.Require().NoError(err)
```

robert-zaremba commented on crypto/keyring/output_test.go:
```diff
@@ -17,14 +17,17 @@ func TestBech32KeysOutput(t *testing.T) {
 	require.NoError(t, err)
 	require.Equal(t, expectedOutput, out)
-	require.Equal(t, `{Name:multisig Type:multi Address:cosmos1nf8lf6n4wa43rzmdzwe6hkrnw5guekhqt595cw PubKey:{"@type":"/cosmos.crypto.multisig.LegacyAminoPubKey","threshold":1,"public_keys":[{"@type":"/cosmos.crypto.secp256k1.PubKey","key":"AurroA7jvfPd1AadmmOvWM2rJSwipXfRf8yD6pLbA2DJ"}]} Mnemonic:}`, fmt.Sprintf("%+v", out))
+	require.Equal(t, "{Name:multisig Type:multi Address:cosmos1nf8lf6n4wa43rzmdzwe6hkrnw5guekhqt595cw PubKey:{\"@type\":\"/cosmos.crypto.multisig.LegacyAminoPubKey\",\"threshold\":1,\"public_keys\":[{\"@type\":\"/cosmos.crypto.secp256k1.PubKey\",\"key\":\"AurroA7jvfPd1AadmmOvWM2rJSwipXfRf8yD6pLbA2DJ\"}]} Mnemonic:}", fmt.Sprintf("%+v", out))
```
why changing this line? using \`...\` makes it easier to read.

cyberbono3 replied:
Otherwise test does not pass since I updated `MkAccKeyOutput`. Should I explicitly use `...` or leave it as it is?

robert-zaremba commented on crypto/ledger/ledger_mock.go:
```diff
@@ -8,11 +8,11 @@ import (
 	"github.com/btcsuite/btcd/btcec"
 	"github.com/pkg/errors"
 
+	"github.com/cosmos/go-bip39"
+
 	secp256k1 "github.com/tendermint/btcd/btcec"
```
```suggestion
	"github.com/cosmos/go-bip39"
	secp256k1 "github.com/tendermint/btcd/btcec"
```

robert-zaremba reviewed (commented):
Good job. There are few corrections to do:
* improve tests
* reduce dupliacted code
* improve go doc comments

cyberbono3 commented on proto/cosmos/crypto/keyring/v1/types.proto:
```diff
@@ -0,0 +1,47 @@
+syntax = "proto3";
+package cosmos.crypto.keyring.v1;
+
```
Sure!

cyberbono3 commented on client/context.go:
```diff
@@ -323,27 +323,32 @@ func GetFromFields(kr keyring.Keyring, from string, genOnly bool) (sdk.AccAddres
 	}
 
-	var info keyring.Info
+	k := new(keyring.Record) // nolint
```
if we dont put `nolint`, linter outputs `client/context.go:326:2            staticcheck  SA4006: this value of `k` is never used` which is not true.

robert-zaremba replied:
the `staticcheck` is right - it says the VALUE is not used. So the proper statement should be:
```
var k *keyring.Record`
```

cyberbono3 commented on client/context.go:
```diff
@@ -323,27 +323,32 @@ func GetFromFields(kr keyring.Keyring, from string, genOnly bool) (sdk.AccAddres
 	}
 
-	return keyring.New(sdk.KeyringServiceName(), backend, ctx.KeyringDir, ctx.Input, ctx.KeyringOptions...)
+	return keyring.New(sdk.KeyringServiceName(), backend, ctx.KeyringDir, ctx.Input, ctx.Codec, ctx.KeyringOptions...)
```
Good idea. I like it.

cyberbono3 commented on client/keys/import_test.go:
```diff
@@ -80,12 +82,15 @@ HbP+c6JmeJy9JXe2rbbF1QtCX1gLqGcDQPBXiCtFvP7/8wTZtVOPj8vREzhZ9ElO
+				WithCodec(cdc)
 			ctx := context.WithValue(context.Background(), client.ClientContextKey, &clientCtx)
 
 			require.NoError(t, err)
```
Sure

cyberbono3 commented on client/tx/tx_test.go:
```diff
@@ -134,24 +134,27 @@ func TestBuildUnsignedTx(t *testing.T) {
 	path := hd.CreateHDPath(118, 0, 0).String()
-	kr, err := keyring.New(t.Name(), "test", t.TempDir(), nil)
+	encCfg := simapp.MakeTestEncodingConfig()
+	kb, err := keyring.New(t.Name(), "test", t.TempDir(), nil, encCfg.Codec)
```
let us stick with `kb` since it is used widely in tests.

robert-zaremba commented on client/context.go:
```diff
@@ -323,27 +323,32 @@ func GetFromFields(kr keyring.Keyring, from string, genOnly bool) (sdk.AccAddres
 	}
 
-	var info keyring.Info
+	k := new(keyring.Record) // nolint
```
```suggestion
	var k *keyring.Record`
```

cyberbono3 commented on crypto/keyring/keyring_test.go:
```diff
@@ -29,69 +32,125 @@ func init() {
+	return true
+}
+
+// getIndices is used in TestInMemoryKeyManagement, TestAltKeyring_List and TestKeyManagementKeyRing
```
sure

cyberbono3 commented:
> Do we need to export NewLegacyMultiInfo? Maybe we can remove it?

We have to export it since It is used in `client/keys` package.

cyberbono3 commented on crypto/keyring/legacy_info.go:
```diff
@@ -261,13 +241,33 @@ func unmarshalInfo(bz []byte) (info Info, err error) {
 	return
 }
+
+// exportPrivateKeyFromLegacyInfo exports a private key from LegacyInfo
+func exportPrivateKeyFromLegacyInfo(info LegacyInfo) (cryptotypes.PrivKey, error) {
```
Done

cyberbono3 commented on crypto/keyring/keyring.go:
```diff
@@ -775,65 +766,69 @@ func newRealPrompt(dir string, buf io.Reader) func(string) (string, error) {
-			return true, nil // address lookup succeeds - info exists
-		}
-		return false, nil
+// different thant ErrKeyNotFound
```
Done

cyberbono3 commented on crypto/keyring/keyring.go:
```diff
@@ -876,31 +877,36 @@ func (ks keystore) MigrateAll() (bool, error) {
 
 // migrate converts keyring.Item from amino to proto serialization format.
-func (ks keystore) migrate(key string) (keyring.Item, bool, error) {
+func (ks keystore) migrate(key string) (*Record, bool, error) {
```
From my perspective, it is more efficient to return `Record,bool,error`. In this case we avoid duplicated invocation of `ks.protoUnmarshalRecord`.

robert-zaremba replied:
OK

cyberbono3 commented on crypto/keyring/codec.go:
```diff
@@ -11,11 +11,13 @@ func init() {
+// rename to LegacyInfo
 func RegisterLegacyAminoCodec(cdc *codec.LegacyAmino) {
-	cdc.RegisterInterface((*Info)(nil), nil)
+	cdc.RegisterInterface((*LegacyInfo)(nil), nil)
```
Nope, cause `func NewLegacyMultiInfo(name string, pub cryptotypes.PubKey) (LegacyInfo, error) {` is used in `client/keys` package. According to linter rule, `LegacyInfo` return type should be public since `NewLegacyMultiInfo` is public too.

cyberbono3 commented on crypto/keyring/record.go:
```diff
@@ -0,0 +1,147 @@
+
+func newMultiRecordItem(re *Record_Multi) *Record_Multi_ {
+	return &Record_Multi_{re}
+}
```
nope, just removed other `new*` funcs

cyberbono3 commented on crypto/keyring/record_test.go:
```diff
@@ -0,0 +1,150 @@
+	cryptotypes "github.com/cosmos/cosmos-sdk/crypto/types"
+)
+
+func TestOfflineRecordMarshaling(t *testing.T) {
```
sounds good.

cyberbono3 commented on server/rosetta/converter.go:
```diff
@@ -315,6 +315,7 @@ func (c converter) Tx(rawTx tmtypes.Tx, txResult *abci.ResponseDeliverTx) (*rose
 	}, nil
 }
 
+//nolint
```
if i remove this `nolint`, linter outputs an issue: `server/rosetta/converter.go:319:2  prealloc  Consider preallocating `ops``

cyberbono3 commented on client/keys/migrate_test.go:
```diff
@@ -3,38 +3,128 @@ package keys
+	SetItem(item design99keyring.Item) error
+}
+
+func Test_runMigrateCmdLegacyInfo(t *testing.T) {
```
Sure

cyberbono3 commented on client/keys/migrate_test.go:
```diff
@@ -2,41 +2,97 @@ package keys
-	require.NoError(t, copy.Copy("testdata", kbHome))
 
-	cmd := MigrateCommand()
+func (s *MigrateTestSuite) Test_runListAndShowCmd() {
```
Should I add more CLI tests here?

robert-zaremba replied:
yes, the  name is good.

We should have a unit test which works directly on the structures without going through the command.

cyberbono3 replied:
You mean [these tests](https://github.com/cosmos/cosmos-sdk/blob/cyberbono3/keyring-migration/crypto/keyring/migration_test.go)?

cyberbono3 replied:
I added Migrate CLI command [tests](https://github.com/cosmos/cosmos-sdk/blob/cyberbono3/keyring-migration/client/keys/migrate_test.go#L93-L150) as well. Does it sound good or you feel I have to improve testing coverage?

robert-zaremba commented on client/keys/migrate_test.go:
```diff
@@ -2,41 +2,97 @@ package keys
+func (s *MigrateTestSuite) SetupSuite() {
+	s.dir = s.T().TempDir()
+	s.cdc = simapp.MakeTestEncodingConfig().Codec
+	s.n1 = "cosmos"
```
`n1` without local context it's too cryptic (1-2 letter names are good for well known abbreviation, or in a local context, like `i := row.Index`, but not otherwise).

robert-zaremba reviewed (commented):
let's add unit test for migration.
Also please include keyring tests into  https://github.com/cosmos/cosmos-sdk/issues/9840

cyberbono3 commented on testutil/network/network.go:
```diff
@@ -338,8 +338,7 @@ func New(l Logger, baseDir string, cfg Config) (*Network, error) {
 
-		encCfg := simapp.MakeTestEncodingConfig()
-		kb, err := keyring.New(sdk.KeyringServiceName(), keyring.BackendTest, clientDir, buf, encCfg.Codec, cfg.KeyringOptions...)
+		kb, err := keyring.New(sdk.KeyringServiceName(), keyring.BackendTest, clientDir, buf, cfg.Codec, cfg.KeyringOptions...)
```
Following @amaury comment, I get rid of `encCfg := simapp.MakeTestEncodingConfig()` here

cyberbono3 commented on client/keys/migrate_test.go:
```diff
@@ -92,6 +93,65 @@ func (s *MigrateTestSuite) Test_runListAndShowCmd() {
 	s.Require().NoError(cmd.ExecuteContext(ctx))
 }
 
+func (s *MigrateTestSuite)  Test_runMigrateCmdRecord() {
```
added a couple of migrate CLI tests

cyberbono3 commented on client/keys/rename_test.go:
```diff
@@ -39,15 +41,15 @@ func Test_runRenameCmd(t *testing.T) {
 	clientCtx := client.Context{}.
 		WithKeyringDir(kbHome).
-		WithKeyring(kb)
+		WithCodec(cdc)
```
Since old test logic yield an [error](https://play.golang.org/p/DijWsX0iCEN), I would prefer to keep `withCodec(cdc)` . In many cases we apply ` flags.FlagKeyringBackend` on `cmd`. Hence, `NewKeyringFromBackend(clientCtx, keyringBackend)` is used to create a `keyring` which uses `ctx.Codec` under the hood.

cyberbono3 commented on client/keys/migrate_test.go:
```diff
@@ -2,41 +2,97 @@ package keys
-	mockIn, mockOut := testutil.ApplyMockIO(cmd)
+
+	mockIn := testutil.ApplyMockIODiscardOutErr(cmd)
+	kb, err := keyring.New(s.n1, keyring.BackendTest, s.dir, mockIn, s.cdc)
+	s.Require().NoError(err)
+
+	setter, ok := kb.(setter)
+	s.Require().True(ok)
+	s.Require().NoError(setter.SetItem(item))
+
+	clientCtx := client.Context{}.WithKeyring(kb)
+	ctx := context.WithValue(context.Background(), client.ClientContextKey, &clientCtx)
```
The reason I manually create a keyring here is to invoke `SetItem(item)` on keyring  before I put keyring on `clientCtx`. On the other hand, if `client.Context` automatically creates a keyring under the hood, there would be no 'item` on keyring. What do you think, @robert-zaremba ?

cyberbono3 commented on client/keys/migrate.go:
```diff
@@ -1,147 +1,40 @@
-const migratePassphrase = "NOOP_PASSPHRASE"
-
 // MigrateCommand migrates key information from legacy keybase to OS secret store.
 func MigrateCommand() *cobra.Command {
```
yes, I just added a couple of tests for this command

cyberbono3 commented on client/keys/migrate_test.go:
```diff
@@ -2,41 +2,97 @@ package keys
 import (
 	"context"
-	"fmt"
+     "fmt"
```
I dont see this formatting error anymore

cyberbono3 commented on client/keys/migrate_test.go:
```diff
@@ -27,23 +27,21 @@ type setter interface {
 	s.dir = s.T().TempDir()
 	s.cdc = simapp.MakeTestEncodingConfig().Codec
-	s.n1 = "cosmos"
+	s.appName = "cosmos"
```
@robert-zaremba Following your suggestion, I renamed `s.n1` to `s.appName`. Does it sound good to you?

cyberbono3 commented on crypto/keyring/legacy_info.go:
```diff
@@ -261,13 +241,33 @@ func unmarshalInfo(bz []byte) (info Info, err error) {
+		}
+
+		return priv, nil
+	// case legacyLedgerInfo, legacyOfflineInfo, LegacyMultiInfo:
```
@robert-zaremba  I am thinking to set up `case legacyLedgerInfo, legacyOfflineInfo, LegacyMultiInfo`  and  cause panic at `default` case. Do you think it is a good idea?

cyberbono3 replied:
cc @robert-zaremba

robert-zaremba replied:
Error is better than panic.

cyberbono3 commented:
### manual testing

- git checkout checkout master
- make build
- ./build/simd keys add robert --keyring-backend test
```
- name: robert
  type: local
  address: cosmos1u0nspyf36jq3qhr2t8yy4n3pcm96jy80ls46le
  pubkey: '{"@type":"/cosmos.crypto.secp256k1.PubKey","key":"Aot0EHlQuK19653MUvovFESajRO/zVNxmF8toQnfUa5X"}'
  mnemonic: ""
```
- ./build/simd keys add andrei --keyring-backend test
```
- name: andrei
  type: local
  address: cosmos1afvahvez4n3jgwhlpf7lyjypw35jgupk2nr76j
  pubkey: '{"@type":"/cosmos.crypto.secp256k1.PubKey","key":"ArUdTIoe0sviJGqpJvKlvjOPBc9nlyqdeToalMS2spff"}'
  mnemonic: ""
```
- ./build/simd keys list --keyring-backend test
```
- name: andrei
  type: local
  address: cosmos1afvahvez4n3jgwhlpf7lyjypw35jgupk2nr76j
  pubkey: '{"@type":"/cosmos.crypto.secp256k1.PubKey","key":"ArUdTIoe0sviJGqpJvKlvjOPBc9nlyqdeToalMS2spff"}'
  mnemonic: ""
- name: robert
  type: local
  address: cosmos1u0nspyf36jq3qhr2t8yy4n3pcm96jy80ls46le
  pubkey: '{"@type":"/cosmos.crypto.secp256k1.PubKey","key":"Aot0EHlQuK19653MUvovFESajRO/zVNxmF8toQnfUa5X"}'
  mnemonic: ""
```

- git checkout cyberbono3/keyring-migration

- ./build/simd keys show andrei robert --keyring-backend test

```
- name: multi
  type: multi
  address: cosmos1vfhvvx49u3dl8q27st2hk6k2ff2kts7a04j756
  pubkey: '{"@type":"/cosmos.crypto.multisig.LegacyAminoPubKey","threshold":1,"public_keys":[{"@type":"/cosmos.crypto.secp256k1.PubKey","key":"ArUdTIoe0sviJGqpJvKlvjOPBc9nlyqdeToalMS2spff"},{"@type":"/cosmos.crypto.secp256k1.PubKey","key":"Aot0EHlQuK19653MUvovFESajRO/zVNxmF8toQnfUa5X"}]}'
  mnemonic: ""
```

- ./build/simd keys delete andrei --keyring-backend test
```
Key reference will be deleted. Continue? [y/N]: y
Key deleted forever (uh oh!)
```

- ./build/simd keys show robert --keyring-backend test
```
- name: robert
  type: local
  address: cosmos1u0nspyf36jq3qhr2t8yy4n3pcm96jy80ls46le
  pubkey: '{"@type":"/cosmos.crypto.secp256k1.PubKey","key":"Aot0EHlQuK19653MUvovFESajRO/zVNxmF8toQnfUa5X"}'
  mnemonic: ""
```

cyberbono3 commented:
Manual testing plan:
- [ ] show
- [ ] migrate
- [ ] list

cyberbono3 commented:
### Add keys and run list commands

- git checkout checkout master

- make build

- ./build/simd keys add robert --keyring-backend test
```
- name: robert
  type: local
  address: cosmos1wwearxffy35lvtf4cyguzs0smkl3vql82q37ge
  pubkey: '{"@type":"/cosmos.crypto.secp256k1.PubKey","key":"A0hErAnskuPYDxZb7+mNYX5Jw49XAO/i/mPxUEAXOArB"}'
  mnemonic: ""
```

- ./build/simd keys add andrei --keyring-backend test
```
- name: andrei
  type: local
  address: cosmos10w9j22fzvlg7cp8v92qyecavvwc2lmp3lrsaep
  pubkey: '{"@type":"/cosmos.crypto.secp256k1.PubKey","key":"Al1LmBLld/3n81ntEsLU1FpW6j476ZrFd2YLfzPD+ZNc"}'
  mnemonic: ""
```

- git checkout cyberbono3/keyring-migration

- make build

//checking if keyname is in DB 

- ./build/simd keys add robert --keyring-backend test
```
override the existing name robert [y/N]: ^C
```

- ./build/simd keys add andrei --keyring-backend test
```
override the existing name andrei [y/N]: N
```

- ./build/simd keys list --list-names  --keyring-backend test
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x17b8d99]

goroutine 1 [running]:
github.com/cosmos/cosmos-sdk/client/keys.runListCmd(0xc001114780, 0xc001110900, 0x0, 0x3, 0x0, 0x0)
        github.com/cosmos/cosmos-sdk/client/keys/list.go:42 +0x159
github.com/spf13/cobra.(*Command).execute(0xc001114780, 0xc0011108d0, 0x3, 0x3, 0xc001114780, 0xc0011108d0)
        github.com/spf13/cobra@v1.2.1/command.go:856 +0x47c
github.com/spf13/cobra.(*Command).ExecuteC(0xc0003b1680, 0x0, 0x0, 0xc000f51bf0)
        github.com/spf13/cobra@v1.2.1/command.go:974 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
        github.com/spf13/cobra@v1.2.1/command.go:902
github.com/spf13/cobra.(*Command).ExecuteContext(...)
        github.com/spf13/cobra@v1.2.1/command.go:895
github.com/cosmos/cosmos-sdk/server/cmd.Execute(0xc0003b1680, 0xc000f51bf0, 0x10, 0x1fe6460, 0xc001044760)
        github.com/cosmos/cosmos-sdk/server/cmd/execute.go:36 +0x285
main.main()
        github.com/cosmos/cosmos-sdk/simapp/simd/main.go:15 +0x45
```

- ./build/simd keys list --keyring-backend test
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0xeaab1c]

goroutine 1 [running]:
github.com/cosmos/cosmos-sdk/crypto/keyring.(*Record).GetPubKey(...)
        github.com/cosmos/cosmos-sdk/crypto/keyring/record.go:64
github.com/cosmos/cosmos-sdk/crypto/keyring.MkAccKeyOutput(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        github.com/cosmos/cosmos-sdk/crypto/keyring/output.go:67 +0x5c
github.com/cosmos/cosmos-sdk/crypto/keyring.MkAccKeysOutput(0xc00105e690, 0xd, 0xd, 0x1b8809f, 0xa, 0x1b80c54, 0x4, 0x1ddd7a8)
        github.com/cosmos/cosmos-sdk/crypto/keyring/output.go:82 +0xbc
github.com/cosmos/cosmos-sdk/client/keys.printKeyringRecords(0x1f91980, 0xc000010018, 0xc00105e690, 0xd, 0xd, 0xc001140918, 0x4)
        github.com/cosmos/cosmos-sdk/client/keys/utils.go:41 +0x4c
github.com/cosmos/cosmos-sdk/client/keys.runListCmd(0xc001132780, 0xc0010e0b80, 0x0, 0x2, 0x0, 0x0)
        github.com/cosmos/cosmos-sdk/client/keys/list.go:37 +0x273
github.com/spf13/cobra.(*Command).execute(0xc001132780, 0xc0010e0b60, 0x2, 0x2, 0xc001132780, 0xc0010e0b60)
        github.com/spf13/cobra@v1.2.1/command.go:856 +0x47c
github.com/spf13/cobra.(*Command).ExecuteC(0xc000dcb680, 0x0, 0x0, 0xc000e6fd40)
        github.com/spf13/cobra@v1.2.1/command.go:974 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
        github.com/spf13/cobra@v1.2.1/command.go:902
github.com/spf13/cobra.(*Command).ExecuteContext(...)
        github.com/spf13/cobra@v1.2.1/command.go:895
github.com/cosmos/cosmos-sdk/server/cmd.Execute(0xc000dcb680, 0xc000e6fd40, 0x10, 0x1fe6460, 0xc0010438c0)
        github.com/cosmos/cosmos-sdk/server/cmd/execute.go:36 +0x285
main.main()
        github.com/cosmos/cosmos-sdk/simapp/simd/main.go:15 +0x45
```
**TODO** add tests for such scenarios and find out the reasons of it.

cyberbono3 commented:
### Migrate command manual testing

- git checkout master

- make build

- ./build/simd keys add andrei --keyring-backend test
```
- name: andrei
  type: local
  address: cosmos1t62uk98ypv9re3u0rfqf0m6fjuh0a428dtw8nc
  pubkey: '{"@type":"/cosmos.crypto.secp256k1.PubKey","key":"AiWLdCNgcSV+wwhlVIF+rYYryy+Jp7crUbWTxmZHqcvl"}'
  mnemonic: ""
```
- ./build/simd keys add robert --keyring-backend test
```
- name: robert
  type: local
  address: cosmos1yx5gtfczr594pkrz0e4melg6x7ywhc2wmc5lgy
  pubkey: '{"@type":"/cosmos.crypto.secp256k1.PubKey","key":"A53BMi2xGHODqkbFhZBCs+2lXa6kkpRrkl0EdOgR53i4"}'
  mnemonic: ""
```

- git checkout cyberbono3/keyring-migration

- make build

-  ./build/simd keys migrate
```
Keys migration has been successfully executed
```
- runMigrateCmd code:
```
func runMigrateCmd(cmd *cobra.Command, _ []string) error {
	clientCtx, err := client.GetClientQueryContext(cmd)
	if err != nil {
		return err
	}

	if _, err = clientCtx.Keyring.MigrateAll(); err != nil {
		return err
	}
	
	cmd.Println("Keys migration has been successfully executed")
	return nil
}
```

cyberbono3 commented:
@robert-zaremba  Should i explicitly run `make build` after checking out `cyberbono3/keyring-migration` branch in show manual [testing](https://github.com/cosmos/cosmos-sdk/pull/9695#issuecomment-914239360)? In this case I get the following error:
```
./build/simd keys show andrei robert --keyring-backend test
Error: andrei is not a valid name or address: decoding bech32 failed: invalid bech32 string length 6
```
Please advice.

cyberbono3 commented on client/keys/list.go:
```diff
@@ -39,6 +39,10 @@ func runListCmd(cmd *cobra.Command, _ []string) error {
 	}
 
 	for _, r := range records {
+		if r == nil {
```
Since I set capacity on `Record` slice [here](https://github.com/cosmos/cosmos-sdk/blob/cyberbono3/keyring-migration/crypto/keyring/keyring.go#L498), some items in `records` are nil. Hence, I have to print out the name of non empty `Record` at L46.

robert-zaremba replied:
Please don't do that. It's error prone that we return `nil` keys in `Keyring.List()`

cyberbono3 replied:
Sure! `Keyring.List()` does not return nil items.

cyberbono3 commented on client/keys/utils.go:
```diff
@@ -60,9 +60,16 @@ func printKeyringRecords(w io.Writer, records []*cryptokeyring.Record, output st
-	if err != nil {
-		panic(err)
+	for _, ko := range kos {
+		if ko == (cryptokeyring.KeyOutput{}) {
```
See [comment](https://github.com/cosmos/cosmos-sdk/pull/9695/commits/cf3e898c627c5fd054e66b955d9d132415b95ecb#r705228168). Hence,  I have to print out non empty `KeyOutput` structs only.

cyberbono3 commented on crypto/keyring/output.go:
```diff
@@ -78,8 +78,12 @@ func MkAccKeyOutput(k *Record) (KeyOutput, error) {
-	for i, re := range records {
-		kos[i], err = MkAccKeyOutput(re)
+	for i, r := range records {
+		if r == nil {
```
I am not sure if `r== nil` check is required in this case.

cyberbono3 commented:
Trying to resolve this linter issues:
```
 Error: directive `// nolint: staticcheck` is unused for linter "staticcheck" (nolintlint)
  Error: directive `// nolint: staticcheck` is unused for linter "staticcheck" (nolintlint)
  Error: directive `// nolint: staticcheck` is unused for linter "staticcheck" (nolintlint)
  Error: directive `// nolint: staticcheck` is unused for linter "staticcheck" (nolintlint)
  Error: directive `// nolint: staticcheck` is unused for linter "staticcheck" (nolintlint)
  Error: directive `// nolint: staticcheck` is unused for linter "staticcheck" (nolintlint)
```
I updated `.golangci.yml` locally and set :
```
nolintlint:
    allow-unused: true
    allow-leading-space: true
    require-explanation: true
    require-specific: true
```
I ran `make lint`  one more time locally  and linter showed me the files where these errors have been occurred. 

1. server/grpc/gogoreflection/fix_registration.go:12:2     nolintlint  directive `// nolint: staticcheck` should provide explanation such as `// nolint: staticcheck // this is why`
2. server/grpc/gogoreflection/fix_registration.go:89:2     nolintlint  directive `// nolint: staticcheck` should provide explanation such as `// nolint: staticcheck // this is why`
3. server/grpc/gogoreflection/fix_registration.go:98:2     nolintlint  directive `// nolint: staticcheck` should provide explanation such as `// nolint: staticcheck // this is why`
4. server/grpc/gogoreflection/fix_registration.go:110:2    nolintlint  directive `// nolint: staticcheck` should provide explanation such as `// nolint: staticcheck // this is why`
5. server/grpc/gogoreflection/fix_registration.go:136:2    nolintlint  directive `// nolint: staticcheck` should provide explanation such as `// nolint: staticcheck // this is why`
6. server/grpc/gogoreflection/serverreflection.go:50:2     nolintlint  directive `// nolint: staticcheck` should provide explanation such as `// nolint: staticcheck // this is why`
7. crypto/codec/amino.go:29:43                             nolintlint  directive `//nolint:staticcheck` should provide explanation such as `//nolint:staticcheck // this is why`


So I removed all 6 `// nolint: staticcheck` comments in files listed above  and ran `make lint` again. There were no more `static check`  lint issues found.
I have pushed modified files and linter at Github Actions outputted the following  6 issues:
```
 Error: SA1019: proto.FileDescriptor is deprecated: Use protoregistry.GlobalFiles.FindFileByPath instead.  (staticcheck)
 Error: SA1019: proto.MessageType is deprecated: Use protoregistry.GlobalTypes.FindMessageByName instead.  (staticcheck)
 Error: SA1019: proto.RegisteredExtensions is deprecated: Use protoregistry.GlobalTypes.RangeExtensionsByMessage instead.  (staticcheck)
 Error: SA1019: proto.RegisteredExtensions is deprecated: Use protoregistry.GlobalTypes.RangeExtensionsByMessage instead.  (staticcheck)
 Error: SA1019: package github.com/golang/protobuf/proto is deprecated: Use the "google.golang.org/protobuf/proto" package instead.  (staticcheck)
 Error: SA1019: package github.com/golang/protobuf/proto is deprecated: Use the "google.golang.org/protobuf/proto" package instead.  (staticcheck)
```
I am a bit confused whether I have to resolve these ` Error: SA1019` errors.

Q: How can I fix these issues?  i googled, but found no clear solution. Perhaps, this is [one] (https://github.com/golangci/golangci-lint/issues/741#issuecomment-596236953), is not it?

cyberbono3 commented on client/keys/migrate.go:
```diff
@@ -1,147 +1,43 @@
-	cmd.PrintErrln("Migration complete.")
-
-	return err
+	cmd.Println("Keys migration has been successfully executed")
```
I made `cmd.Println` only for testing purposes to make sure `keys migrate` CLI actually works as expected. Should I leave this line or remove it?

robert-zaremba replied:
This one you can leave. But you should remove most of other printlns.

cyberbono3 replied:
what other printlns do you mean?

cyberbono3 commented on crypto/keyring/keyring_test.go:
```diff
@@ -1162,13 +1356,13 @@ func TestRenameKey(t *testing.T) {
 				require.NoError(t, err)
-				requireEqualRenamedKey(t, newInfo, oldKeyInfo) // oldinfo and newinfo should be the same
-				oldKeyInfo, err = kr.Key(oldKeyUID)            // old key should be gone from keyring
+				requireEqualRecord(t, newRecord, oldKeyRecord, false) // oldKeyRecord and newRecord should be the same except name
```
After running `kr.Rename(oldKeyUID, newKeyUID)` , all fields of `newRecord` and `oldKeyRecord` should match except names (`newRecord.Name = "new"`, `oldKeyRecord.Name = "old"`). Are you okay with that, @robert-zaremba ?

robert-zaremba replied:
why not using the `requireEqualRenamedKey`?

cyberbono3 replied:
OK

cyberbono3 commented on client/keys/utils.go:
```diff
@@ -71,10 +59,17 @@ func printInfos(w io.Writer, infos []cryptokeyring.Info, output string) {
+			continue
+		}
+
+		out, err := yaml.Marshal(ko)
```
If introduces a linter issue if you use `yaml.Marshal(&ko)` inside for loop. That's why I decided loop over slice of KeyOutput pointers.

robert-zaremba commented on client/keys/list_test.go:
```diff
@@ -46,10 +49,10 @@ func Test_runListCmd(t *testing.T) {
-		{"keybase: empty", kbHome1, false},
-		{"keybase: w/key", kbHome2, false},
+		{"keybase: empty", kbHome1, keyring.ErrNoKeysAvailable},
+		{"keybase: w/key", kbHome2, keyring.ErrNoKeysAvailable},
```
Why do you want to return an error if there are no keys?

cyberbono3 replied:
`keyring.List()` returns no error if there are no keys

cyberbono3 replied:
you are right. Now i return no error when there are no keys

robert-zaremba commented on crypto/keyring/legacy_info.go:
```diff
@@ -261,13 +241,33 @@ func unmarshalInfo(bz []byte) (info Info, err error) {
+		return priv, nil
+	// case legacyLedgerInfo, legacyOfflineInfo, LegacyMultiInfo:
+	default:
+		return nil, errors.New("only works on local private keys")
```
In the error message, please add the `legacyLocalInfo`, eg:
```
fmt.Errorf("only works on local private keys, provided %s", linfo)
```

cyberbono3 replied:
See [Comment](https://github.com/cosmos/cosmos-sdk/pull/9695#discussion_r710016546) what other printlns do you mean, @robert-zaremba ?

robert-zaremba reviewed (approved):
Pre approving
- [x] `Keyring.List()` should not return nil keys
- [ ]  remove not needed printlns

Thanks for your comments Andrei - they help with the final review.

cyberbono3 commented:
See [Comment](https://github.com/cosmos/cosmos-sdk/pull/9695#discussion_r710016546) what other printlns do you mean, @robert-zaremba ?

cyberbono3 commented on crypto/keyring/keyring_test.go:
```diff
@@ -1211,32 +1360,35 @@ func TestRenameKey(t *testing.T) {
-	require.Equal(t, newKey.GetType(), oldKey.GetType())
-}
+func requireEqualRenamedKey(t *testing.T, key *Record, mnemonic *Record, nameMatch bool) {
+	if nameMatch {
```
I added 'nameMatch'  argument to ` requireEqualRenamedKey` as in some cases i dont need name check ( e.g. at L1319). If it is set to true, it  checks if `key.Name` and `mnemonic.Name` match as you can see below. I hope it sounds good to you, @robert-zaremba

amaurym commented:
Thanks a lot @cyberbono3 for your work on this PR!

alessio commented:
I hope this work was completed putting *legacy clients first* and ensuring that Cosmos Hub users would not stumble on *any disruption whatsoever*. Else, you will hear people shouting very, very loud with the next release.

As a word of advice, I'd test this *thoroughly* before releasing it.
